pkg/server/commands/list_objects.go

```diff
@@ -95,76 +113,121 @@ func (q *ListObjectsQuery) evaluate(
-			zap.String("object_type", targetObjectType),
-		)
-	}
-
-	containsIntersection, _ := typesys.RelationInvolvesIntersection(targetObjectType, targetRelation)
-	containsExclusion, _ := typesys.RelationInvolvesExclusion(targetObjectType, targetRelation)
-
-	// ConnectedObjects currently only supports models that do not include intersection and exclusion,
-	// and the model must include type info for ConnectedObjects to work.
-	if !containsIntersection && !containsExclusion && hasTypeInfo {
 		userObj, userRel := tuple.SplitObjectRelation(req.GetUser())
-
 		userObjType, userObjID := tuple.SplitObject(userObj)
 
-		var targetUserRef isUserRef
-		targetUserRef = &UserRefObject{
+		var sourceUserRef isUserRef
+		sourceUserRef = &UserRefObject{
 			Object: &openfgapb.Object{
 				Type: userObjType,
 				Id:   userObjID,
 			},
 		}
 
 		if tuple.IsTypedWildcard(userObj) {
-			targetUserRef = &UserRefTypedWildcard{Type: tuple.GetType(userObj)}
+			sourceUserRef = &UserRefTypedWildcard{Type: tuple.GetType(userObj)}
+
 		}
 
 		if userRel != "" {
-			targetUserRef = &UserRefObjectRelation{
+			sourceUserRef = &UserRefObjectRelation{
 				ObjectRelation: &openfgapb.ObjectRelation{
 					Object:   userObj,
 					Relation: userRel,
 				},
 			}
 		}
 
-		handler = func() {
-			defer close(resultsChan)
-
-			span.SetAttributes(attribute.Bool(listObjectsOptimizedKey, true))
+		connectedObjectsResChan := make(chan *ConnectedObjectsResult, 1)
+		var objectsFound = new(uint32)
 
-			connectObjCmd := &ConnectedObjectsCommand{
-				Datastore:        q.Datastore,
-				ResolveNodeLimit: q.ResolveNodeLimit,
-				Limit:            maxResults,
-			}
+		connectedObjectsCmd := &ConnectedObjectsCommand{
+			Datastore:        q.Datastore,
+			Typesystem:       typesys,
+			ResolveNodeLimit: q.ResolveNodeLimit,
+			Limit:            maxResults,
+		}
 
-			err = connectObjCmd.StreamedConnectedObjects(ctx, &ConnectedObjectsRequest{
+		go func() {
+			err = connectedObjectsCmd.StreamedConnectedObjects(ctx, &ConnectedObjectsRequest{
 				StoreID:          req.GetStoreId(),
 				Typesystem:       typesys,
 				ObjectType:       targetObjectType,
 				Relation:         targetRelation,
-				User:             targetUserRef,
+				User:             sourceUserRef,
 				ContextualTuples: req.GetContextualTuples().GetTupleKeys(),
-			}, resultsChan)
+			}, connectedObjectsResChan)
+			if err != nil {
+				resultsChan <- ListObjectsResult{Err: err}
+			}
+
+			close(connectedObjectsResChan)
+		}()
+
+		checkResolver := graph.NewLocalChecker(
+			q.Datastore,
+			q.CheckConcurrencyLimit,
+		)
+
+		concurrencyLimiterCh := make(chan struct{}, maximumConcurrentChecks)
+
+		wg := sync.WaitGroup{}
+
+		for res := range connectedObjectsResChan {
+
+			if res.ResultStatus == NoFurtherEvalStatus && atomic.AddUint32(objectsFound, 1) <= maxResults {
+				noFurtherEvalRequiredCounter.Inc()
+				resultsChan <- ListObjectsResult{ObjectID: res.Object}
+				continue
+			}
```

Comment:
shouldn't this be 

```go
                         if res.ResultStatus == NoFurtherEvalStatus {
				noFurtherEvalRequiredCounter.Inc()
				if atomic.AddUint32(objectsFound, 1) <= maxResults {
					resultsChan <- ListObjectsResult{ObjectID: res.Object}
				}
				continue
			}
```

Reply:
You should only increment the counter if you yield a result. Your code increments the counter regardless of whether a result is yielded or not.

It may be smart for us to look into returning early though if `atomic.AddUint32(objectsFound, 1) > maxResults` though. The `continue` will move on to the next item even though we won't yield anymore. So that's quite wasteful.

Reply:
Oh I see your point now. The counters will potentially be unbalanced in the event that you've already found maxResults. I'll fix 👍

Reply:
Fixed. Thanks!